src/Linker.DefineFunction.tt

```diff
@@ -90,26 +97,29 @@ foreach (var (hasCaller, resultCount, parameterCount, methodGenerics, delegateTy
 <# GenerateCallbackContent(hasCaller, resultCount, parameterCount); #>
                 };
 
-                var moduleBytes = Encoding.UTF8.GetBytes(module);
-                var nameBytes = Encoding.UTF8.GetBytes(name);
-                fixed (byte* modulePtr = moduleBytes, namePtr = nameBytes)
+                using var moduleBytes = MemoryPool<byte>.Shared.Rent(Encoding.UTF8.GetByteCount(module));
+                var moduleBytesLength = Encoding.UTF8.GetBytes(module, moduleBytes.Memory.Span);
+                using var modulePin = moduleBytes.Memory.Pin();
+
+                using var nameBytes = MemoryPool<byte>.Shared.Rent(Encoding.UTF8.GetByteCount(name));
+                var nameBytesLength = Encoding.UTF8.GetBytes(name, nameBytes.Memory.Span);
+                using var namePin = nameBytes.Memory.Pin();
+
+                var error = Native.wasmtime_linker_define_func_unchecked(
+                    handle,
+                    (byte*)modulePin.Pointer,
+                    (nuint)moduleBytesLength,
+                    (byte*)namePin.Pointer,
+                    (nuint)nameBytesLength,
+                    funcType,
+                    func,
+                    GCHandle.ToIntPtr(GCHandle.Alloc(func)),
+                    Function.Finalizer
+                );
```

Comment:
> I've rewritten it to use the shared `MemoryPool<byte>` instead of stack allocations.

Thanks!
However, note that, unfortunately, AFAIK using `MemoryPool<T>.Shared.Rent` will always cause an allocation of at least one object (in contrast to `ArrayPool<T>.Shared.Rent`), in this case the `IMemoryOwner<T>`. See these code locations in the runtime:
https://github.com/dotnet/runtime/blob/322f20cda9673a3872902ed3fb2ddfbcff97c007/src/libraries/System.Memory/src/System/Buffers/ArrayMemoryPool.cs#L21
https://github.com/dotnet/runtime/blob/322f20cda9673a3872902ed3fb2ddfbcff97c007/src/libraries/System.Memory/src/System/Buffers/ArrayMemoryPool.ArrayMemoryPoolBuffer.cs#L8-L15

I think for best efficiency (completely avoid heap allocations in the most common case, when the `module` and `name` strings are not very long), it might make sense to apply the implementation from the `MultipartContent` example, which uses `stackalloc byte[...]` for strings shorter than the threshold, and uses `ArrayPool<byte>.Shared.Rent(...)` for longer strings (although the code is a bit more verbose).

What do you think?

```suggestion
                const int StackallocThreshold = 256;

                byte[]? moduleBytesBuffer = null;
                var moduleLength = Encoding.UTF8.GetByteCount(module);
                Span<byte> moduleBytes = moduleLength <= StackallocThreshold ? stackalloc byte[moduleLength] : (moduleBytesBuffer = ArrayPool<byte>.Shared.Rent(moduleLength))[..moduleLength];
                Encoding.UTF8.GetBytes(module, moduleBytes);

                byte[]? nameBytesBuffer = null;
                var nameLength = Encoding.UTF8.GetByteCount(name);
                Span<byte> nameBytes = nameLength <= StackallocThreshold ? stackalloc byte[nameLength] : (nameBytesBuffer = ArrayPool<byte>.Shared.Rent(nameLength))[..nameLength];
                Encoding.UTF8.GetBytes(name, nameBytes);

                try
                {
                    fixed (byte* modulePtr = moduleBytes, namePtr = nameBytes)
                    {
                        var error = Native.wasmtime_linker_define_func_unchecked(
                            handle,
                            modulePtr,
                            (nuint)moduleBytes.Length,
                            namePtr,
                            (nuint)nameBytes.Length,
                            funcType,
                            func,
                            GCHandle.ToIntPtr(GCHandle.Alloc(func)),
                            Function.Finalizer
                        );

                        if (error != IntPtr.Zero)
                        {
                            throw WasmtimeException.FromOwnedError(error);
                        }
                    }
                }
                finally
                {
                    if (moduleBytesBuffer is not null)
                    {
                        ArrayPool<byte>.Shared.Return(moduleBytesBuffer);
                    }
                    if (nameBytesBuffer is not null)
                    {
                        ArrayPool<byte>.Shared.Return(nameBytesBuffer);
                    }
                }
```

Reply:
You're absolutely right, not sure how I missed that MemoryPool would allocate that handle 🤦‍♂️

I only used `MemoryPool` instead of `ArrayPool` out of neatness - it looked cleaner to have a `using` instead of wrapping everything in try/finally!